src/mono/mono/mini/interp/transform-simd.c

```diff
@@ -455,6 +468,25 @@ emit_sri_vector128 (TransformData *td, MonoMethod *cmethod, MonoMethodSignature
 			simd_intrins = INTERP_SIMD_INTRINSIC_V128_BITCAST;
 			break;
 		}
+		case SN_AsVector:
+		case SN_AsVector128:
+		case SN_AsVector4: {
+			if (!is_element_type_primitive (csignature->ret) || !is_element_type_primitive (csignature->params [0]))
+				return FALSE;
+
+			MonoClass *ret_class = mono_class_from_mono_type_internal (csignature->ret);
+			int ret_size = mono_class_value_size (ret_class, NULL);
+
+			MonoClass *arg_class = mono_class_from_mono_type_internal (csignature->params [0]);
+			int arg_size = mono_class_value_size (arg_class, NULL);
+
+			if (arg_size == ret_size) {
```

Comment:
What are the scenarios where this check would fail? `Vector<decimal>` or something? Or is this mostly a guard to catch someone adding `AsVector2`/`AsVector3` without updating everything?

Reply:
We have APIs like `Vector128<float> AsVector128(Vector2 vector)` in which case `arg_size == 8` and we don't want to bitcast, we'd need separate handling instead.

This also protects against `Vector<T>` expanding to be larger than 16-bytes in the future, which could happen for `x64` on an `AVX2` capable machine or for `Arm64` on an `SVE` capable machine.

In general we likely want to try to be as exact as possible where we can and in RyuJIT we actually validate every parameter type via the general `getBaseJitTypeAndSizeOfSIMDType` and `isSupportedBaseType` APIs, which allows us to see if something is SIMD or not and then validate it's supported if so. We then bail out of intrinsic importation and fallback to treating it as a normal managed call. We correspondingly track whether a given intrinsic needs to get the base type used for determining the correct importation behavior based on flags such as `HW_Flag_BaseTypeFromFirstArg` or `HW_Flag_BaseTypeFromSecondArg` and the general shape expected of the parameters via similar flags and the "category". -- The general logic is https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsic.cpp#L1365, which flows from a table like https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsiclistxarch.h, and using helpers defined in https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsic.h